### ISSUE TITLE
fatal error: concurrent map writes

### ISSUE THREAD

**Issue by shaotingEri** [openstack-exporter/openstack-exporter#166]
Hi,
When I use this command:
$ nohup ./openstack-exporter --os-client-config=xxxxx/clouds.yaml    --multi-cloud --collect-metric-time &

It always crash after some days.
From the log: fatal error: concurrent map writes, I guess maybe need some lock mechanism when deal with multi clouds.
Below is the error information:

fatal error: concurrent map writes

goroutine 10649726 [running]:
runtime.throw(0xa56c52, 0x15)
        /usr/local/go/src/runtime/panic.go:774 +0x72 fp=0xc000211108 sp=0xc0002110d8 pc=0x42e642
runtime.mapassign_faststr(0x9884c0, 0xc000289c80, 0xa4f9a7, 0x5, 0xb)
        /usr/local/go/src/runtime/map_faststr.go:211 +0x417 fp=0xc000211170 sp=0xc000211108 pc=0x4131a7
github.com/openstack-exporter/openstack-exporter/exporters.NewServiceClient(0xa4f9a7, 0x5, 0xc0002118d8, 0xc000142dc0, 0xa4fe70, 0x6, 0x0, 0x0, 0x0)
        /app/exporters/utils.go:112 +0x2d8 fp=0xc000211830 sp=0xc000211170 pc=0x8ea168
github.com/openstack-exporter/openstack-exporter/exporters.NewExporter(0xa4f9a7, 0x5, 0xa511a8, 0x9, 0xc0008fa091, 0x4, 0xc0001188f0, 0x1, 0x1, 0xa4fe70, ...)
        /app/exporters/exporter.go:193 +0xfb fp=0xc000211958 sp=0xc000211830 pc=0x8db01b
github.com/openstack-exporter/openstack-exporter/exporters.EnableExporter(0xa4f9a7, 0x5, 0xa511a8, 0x9, 0xc0008fa091, 0x4, 0xc0001188f0, 0x1, 0x1, 0xa4fe70, ...)
        /app/exporters/exporter.go:43 +0x104 fp=0xc0002119f8 sp=0xc000211958 pc=0x8d9654
main.probeHandler(0xb1c320, 0xc0001f80e0, 0xc001002100)
        /app/main.go:124 +0x4d7 fp=0xc000211b90 sp=0xc0002119f8 pc=0x902ba7
net/http.HandlerFunc.ServeHTTP(0xa757e0, 0xb1c320, 0xc0001f80e0, 0xc001002100)
        /usr/local/go/src/net/http/server.go:2036 +0x44 fp=0xc000211bb8 sp=0xc000211b90 pc=0x6c2c54
net/http.(*ServeMux).ServeHTTP(0xf02b40, 0xb1c320, 0xc0001f80e0, 0xc001002100)
        /usr/local/go/src/net/http/server.go:2416 +0x1bd fp=0xc000211c18 sp=0xc000211bb8 pc=0x6c4b2d
net/http.serverHandler.ServeHTTP(0xc0001a6000, 0xb1c320, 0xc0001f80e0, 0xc001002100)
        /usr/local/go/src/net/http/server.go:2831 +0xa4 fp=0xc000211c48 sp=0xc000211c18 pc=0x6c60a4
net/http.(*conn).serve(0xc0003f2000, 0xb1e7e0, 0xc000090440)
        /usr/local/go/src/net/http/server.go:1919 +0x875 fp=0xc000211fc8 sp=0xc000211c48 pc=0x6c1a45
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc000211fd0 sp=0xc000211fc8 pc=0x45ac11
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:2957 +0x384

goroutine 1 [IO wait, 665 minutes]:
internal/poll.runtime_pollWait(0x7f780989d080, 0x72, 0x0)
        /usr/local/go/src/runtime/netpoll.go:184 +0x55
internal/poll.(*pollDesc).wait(0xc000120298, 0x72, 0x0, 0x0, 0xa502ce)
        /usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
        /usr/local/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Accept(0xc000120280, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /usr/local/go/src/internal/poll/fd_unix.go:384 +0x1f8
net.(*netFD).accept(0xc000120280, 0xc000063c70, 0x6c7034, 0xc0001a60a0)
        /usr/local/go/src/net/fd_unix.go:238 +0x42
net.(*TCPListener).accept(0xc0001132c0, 0x604bb8ac, 0xc000063c70, 0x47e0a6)
        /usr/local/go/src/net/tcpsock_posix.go:139 +0x32
net.(*TCPListener).Accept(0xc0001132c0, 0xc000063cc0, 0x18, 0xc000000180, 0x6c6564)
        /usr/local/go/src/net/tcpsock.go:261 +0x47
net/http.(*Server).Serve(0xc0001a6000, 0xb1c060, 0xc0001132c0, 0x0, 0x0)
        /usr/local/go/src/net/http/server.go:2925 +0x280
net/http.Serve(...)
        /usr/local/go/src/net/http/server.go:2497
main.main()
        /app/main.go:95 +0x98c

goroutine 21782909 [IO wait]:
internal/poll.runtime_pollWait(0x7f7809820be0, 0x72, 0xffffffffffffffff)
        /usr/local/go/src/runtime/netpoll.go:184 +0x55
internal/poll.(*pollDesc).wait(0xc001067018, 0x72, 0x0, 0x1, 0xffffffffffffffff)
        /usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
        /usr/local/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc001067000, 0xc0021aa3d1, 0x1, 0x1, 0x0, 0x0, 0x0)
        /usr/local/go/src/internal/poll/fd_unix.go:169 +0x1cf
net.(*netFD).Read(0xc001067000, 0xc0021aa3d1, 0x1, 0x1, 0x0, 0x0, 0xc000722a98)
        /usr/local/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc0000b6198, 0xc0021aa3d1, 0x1, 0x1, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/net.go:184 +0x68
net/http.(*connReader).backgroundRead(0xc0021aa3c0)
        /usr/local/go/src/net/http/server.go:688 +0x58
created by net/http.(*connReader).startBackgroundRead
        /usr/local/go/src/net/http/server.go:684 +0xd4

goroutine 8836989 [IO wait]:
internal/poll.runtime_pollWait(0x7f78096675a8, 0x72, 0xffffffffffffffff)
        /usr/local/go/src/runtime/netpoll.go:184 +0x55
internal/poll.(*pollDesc).wait(0xc0012cd818, 0x72, 0x1000, 0x1000, 0xffffffffffffffff)
        /usr/local/go/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
        /usr/local/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc0012cd800, 0xc00197b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /usr/local/go/src/internal/poll/fd_unix.go:169 +0x1cf
net.(*netFD).Read(0xc0012cd800, 0xc00197b000, 0x1000, 0x1000, 0x4692fc, 0xef4360, 0x7f7809738918)
        /usr/local/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc00011a638, 0xc00197b000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/net.go:184 +0x68
net/http.(*connReader).Read(0xc0013fe360, 0xc00197b000, 0x1000, 0x1000, 0x44395c, 0x403b74, 0x203000)
        /usr/local/go/src/net/http/server.go:796 +0xf4
bufio.(*Reader).fill(0xc00135a420)
        /usr/local/go/src/bufio/bufio.go:100 +0x103
bufio.(*Reader).ReadSlice(0xc00135a420, 0xa, 0x28, 0xc0002459a8, 0x40bf36, 0xc000cb8300, 0x100)
        /usr/local/go/src/bufio/bufio.go:359 +0x3d
bufio.(*Reader).ReadLine(0xc00135a420, 0xc0002459b0, 0xc00047a000, 0x7f780bcab1f0, 0x0, 0x40c7a8, 0x30)
        /usr/local/go/src/bufio/bufio.go:388 +0x34
net/textproto.(*Reader).readLineSlice(0xc000ea5560, 0xc000cb8300, 0xc0012cd800, 0x0, 0x0, 0x43043c)
        /usr/local/go/src/net/textproto/reader.go:57 +0x6c
net/textproto.(*Reader).ReadLine(...)
        /usr/local/go/src/net/textproto/reader.go:38
net/http.readRequest(0xc00135a420, 0x0, 0xc000cb8300, 0x0, 0x0)
        /usr/local/go/src/net/http/request.go:1012 +0x92
net/http.(*conn).readRequest(0xc000fc0460, 0xb1e7e0, 0xc0018ac6c0, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/http/server.go:982 +0x15f
net/http.(*conn).serve(0xc000fc0460, 0xb1e7e0, 0xc0018ac6c0)
        /usr/local/go/src/net/http/server.go:1845 +0x6d4
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:2957 +0x384

goroutine 8848363 [runnable]:
github.com/openstack-exporter/openstack-exporter/exporters.(*BaseOpenStackExporter).AddMetric(0xc001395c80, 0xa5252b, 0xc, 0xa752e0, 0x0, 0x0, 0x0, 0xc000247460)
        /app/exporters/exporter.go:165 +0xfc
github.com/openstack-exporter/openstack-exporter/exporters.NewNeutronExporter(0xc000247890, 0x7, 0xa503d1, 0x7)
        /app/exporters/neutron.go:53 +0x225
github.com/openstack-exporter/openstack-exporter/exporters.NewExporter(0xa50433, 0x7, 0xa511a8, 0x9, 0xc00003e111, 0x4, 0xc0001188f0, 0x1, 0x1, 0xa4fe70, ...)
        /app/exporters/exporter.go:214 +0x717
github.com/openstack-exporter/openstack-exporter/exporters.EnableExporter(0xa50433, 0x7, 0xa511a8, 0x9, 0xc00003e111, 0x4, 0xc0001188f0, 0x1, 0x1, 0xa4fe70, ...)
        /app/exporters/exporter.go:43 +0x104
main.probeHandler(0xb1c320, 0xc000362a80, 0xc000896600)
        /app/main.go:124 +0x4d7
net/http.HandlerFunc.ServeHTTP(0xa757e0, 0xb1c320, 0xc000362a80, 0xc000896600)
        /usr/local/go/src/net/http/server.go:2036 +0x44
net/http.(*ServeMux).ServeHTTP(0xf02b40, 0xb1c320, 0xc000362a80, 0xc000896600)
        /usr/local/go/src/net/http/server.go:2416 +0x1bd
net/http.serverHandler.ServeHTTP(0xc0001a6000, 0xb1c320, 0xc000362a80, 0xc000896600)
        /usr/local/go/src/net/http/server.go:2831 +0xa4
net/http.(*conn).serve(0xc000803e00, 0xb1e7e0, 0xc00161d800)
        /usr/local/go/src/net/http/server.go:1919 +0x875
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:2957 +0x384

**Comment by shaotingEri** [comment]
by the way, we had almost 15 clouds, and fetch the data from Prometheus every 10s(by default).

**Comment by niedbalski** [comment]
@shaotingEri are you still experiencing this issue? can you upload a escaped version of your configuration? is this with multi-cloud support?

**Comment by shaotingEri** [comment]
@niedbalski
With the same configuration: clouds.yaml, it OK when use the single mode, but fail with the --multi-cloud.
$ nohup ./openstack-exporter --os-client-config=xxxxx/clouds.yaml c204 &
---> Work well.

$ nohup ./openstack-exporter --os-client-config=xxxxx/clouds.yaml --multi-cloud --collect-metric-time &
----> Crash as this IO error after two or three days.

**Comment by sammic30** [comment]
Hi,

We are also facing the same issue when running with the--multi-cloud.
`fatal error: concurrent map writes

goroutine 1029099 [running]:
runtime.throw(0x9edf55, 0x15)
        /snap/go/7360/src/runtime/panic.go:1117 +0x72 fp=0xc00024f048 sp=0xc00024f018 pc=0x438bf2
runtime.mapassign_faststr(0x940040, 0xc00009da10, 0x9ea2b5, 0xd, 0xb)
        /snap/go/7360/src/runtime/map_faststr.go:291 +0x3d8 fp=0xc00024f0b0 sp=0xc00024f048 pc=0x416298
github.com/openstack-exporter/openstack-exporter/exporters.NewServiceClient(0x9ea2b5, 0xd, 0xc00024f7e0, 0x0, 0x9e7367, 0x6, 0x0, 0x0, 0x0)
        /home/circleci/project/parts/openstack-exporter/build/exporters/utils.go:112 +0x2e7 fp=0xc00024f738 sp=0xc00024f0b0 pc=0x8b6c07
github.com/openstack-exporter/openstack-exporter/exporters.NewExporter(0x9ea2b5, 0xd, 0x9e8651, 0x9, 0xc005631c71, 0x5, 0xc0001827e0, 0x1, 0x1, 0x9e7367, ...)
        /home/circleci/project/parts/openstack-exporter/build/exporters/exporter.go:205 +0x105 fp=0xc00024f860 sp=0xc00024f738 pc=0x8a5485
github.com/openstack-exporter/openstack-exporter/exporters.EnableExporter(0x9ea2b5, 0xd, 0x9e8651, 0x9, 0xc005631c71, 0x5, 0xc0001827e0, 0x1, 0x1, 0x9e7367, ...)
        /home/circleci/project/parts/openstack-exporter/build/exporters/exporter.go:44 +0x110 fp=0xc00024f900 sp=0xc00024f860 pc=0x8a3950
main.probeHandler.func1(0xa8c2a0, 0xc0006e42a0, 0xc0001dac00)
        /home/circleci/project/parts/openstack-exporter/build/main.go:132 +0x64c fp=0xc00024fae8 sp=0xc00024f900 pc=0x8cf16c
net/http.HandlerFunc.ServeHTTP(0xc0001828c0, 0xa8c2a0, 0xc0006e42a0, 0xc0001dac00)
        /snap/go/7360/src/net/http/server.go:2069 +0x44 fp=0xc00024fb10 sp=0xc00024fae8 pc=0x6b36c4
net/http.(*ServeMux).ServeHTTP(0xd9cb60, 0xa8c2a0, 0xc0006e42a0, 0xc0001dac00)
        /snap/go/7360/src/net/http/server.go:2448 +0x1ad fp=0xc00024fb70 sp=0xc00024fb10 pc=0x6b554d
net/http.serverHandler.ServeHTTP(0xc0001f6000, 0xa8c2a0, 0xc0006e42a0, 0xc0001dac00)
        /snap/go/7360/src/net/http/server.go:2887 +0xa3 fp=0xc00024fba0 sp=0xc00024fb70 pc=0x6b6c83
net/http.(*conn).serve(0xc00769e820, 0xa8ded8, 0xc0021e6400)
        /snap/go/7360/src/net/http/server.go:1952 +0x8cd fp=0xc00024ffc8 sp=0xc00024fba0 pc=0x6b21ad
runtime.goexit()
        /snap/go/7360/src/runtime/asm_amd64.s:1371 +0x1 fp=0xc00024ffd0 sp=0xc00024ffc8 pc=0x46e4e1
created by net/http.(*Server).Serve
        /snap/go/7360/src/net/http/server.go:3013 +0x39b

goroutine 1 [IO wait]:
internal/poll.runtime_pollWait(0x7f39d0ba5178, 0x72, 0x0)
        /snap/go/7360/src/runtime/netpoll.go:222 +0x55
internal/poll.(*pollDesc).wait(0xc0001b8318, 0x72, 0x0, 0x0, 0x9e77cd)
        /snap/go/7360/src/internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
        /snap/go/7360/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Accept(0xc0001b8300, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /snap/go/7360/src/internal/poll/fd_unix.go:401 +0x212
net.(*netFD).accept(0xc0001b8300, 0xe70aa5cbe25b43d2, 0x0, 0x0)`

**Comment by vigr** [comment]
I've encountered the issue in production, so here are my findings.

### TL;DR - no synchronization over [`exporters.endpointOpts`](https://github.com/openstack-exporter/openstack-exporter/blob/v1.7.0/exporters/exporter.go#L78) map in 2 cases: initialization and access
Versions affected: all, after commit [`ea2e2f5f`](https://github.com/openstack-exporter/openstack-exporter/commits/ea2e2f5f0a14c299f8543d889d5e5c855d452c53) (2019-12-19)

### Detailed explanation
Given the stack trace and the message `"concurrent map writes`, it's obvious the panic comes from either of 2 functions:
* `runtime.mapassign_faststr`: [here](https://github.com/golang/go/blob/go1.22.5/src/runtime/map_faststr.go#L211-L213) and [here](https://github.com/golang/go/blob/go1.22.5/src/runtime/map_faststr.go#L294-L296)
* `runtime.mapdelete_faststr`: [here](https://github.com/golang/go/blob/go1.22.5/src/runtime/map_faststr.go#L309-L311) and [here](https://github.com/golang/go/blob/go1.22.5/src/runtime/map_faststr.go#L385-L387)

**Conclusion**: there is no synchronization of writes to the map `exporters.endpointOpts`.

Guided by the stacktrace, we see its usage [here](https://github.com/openstack-exporter/openstack-exporter/blob/v1.7.0/exporters/utils.go#L111-L115): a global variable is **initialized** and **accessed** with no apparent sync/lock anywhere.

I'm not familiar with the codebase at all, so I'm guessing 2 possibilities:
1. `endpointOpts` represent a global state, private to package `exporters`. In this case it should be initialized exactly once.
2. `endpointOpts` represents some state, local to a particular instance of `gophercloud.ServiceClient`. In this case it should be stored within the returned client instance, not globally.

I reckon the former is the case.

#### Initialization problem
Suppose 2 goroutines call `NewServiceClient` concerrently: `g1` and `g2`.
1.  `g1` executes line [112](https://github.com/openstack-exporter/openstack-exporter/blob/v1.7.0/exporters/utils.go#L112), sees that `endpointOpts` is `nil` but then gets preempted before it can allocate a new map.
2. `g2` does the same - executes lines [112-115](https://github.com/openstack-exporter/openstack-exporter/blob/v1.7.0/exporters/utils.go#L112-L115): sees `nil`, initializes the map and stores a value in it.
3. `g1` is the scheduled, proceeds to allocate a new map and **re-assigns the global variable**.

So the value stored by `g2` is lost along with the map `g2` instantiated: it's GC-ed.

### Possible solution
It may be tempting to introduce a (global) RW-lock guarding the initialization/access to `endpointOpts`, but then it necessitates addition of `lock/defer unlock` pair to its every usage. A better solution (assuming I guessed the intended use of `endpointOpts`) is to use a [`sync.Map`](https://pkg.go.dev/golang.org/x/sync/syncmap) instead:
* this eliminates the need for synchronized initialization - as per doc, "The zero Map is valid and empty."
* it doesn't require any additional locking
